ADR-033 cosmetic updates (#8676)
cosmos/cosmos-sdk: docs/architecture/adr-033-protobuf-inter-module-comm.md

`SetBalance` method allows the caller to set the balance of any account to anything, bypassing even proper tracking of supply.

There appears to have been some later attempts to implement some semblance of OCAPs using module-level minting, staking
and burning permissions. These permissions allow a module to mint, burn or delegate tokens with reference to the module’s
own account. These permissions are actually stored as a `[]string` array on the `ModuleAccount` type in state.

However, these permissions don’t really do much. They control what modules can be referenced in the `MintCoins`,
`BurnCoins` and `DelegateCoins***` methods, but for one there is no unique object capability token that controls access —
just a simple string. So the `x/upgrade` module could mint tokens for the `x/staking` module simple by calling
`MintCoins(“staking”)`. Furthermore, all modules which have access to these keeper methods, also have access to
`SetBalance` negating any other attempt at OCAPs and breaking even basic object-oriented encapsulation.

## Decision

Based on [ADR-021](./adr-021-protobuf-query-encoding.md) and [ADR-031](./adr-031-msg-service.md), we introduce the

> Inter-Module Communication paradigm for secure module authorization and OCAPS

Reviewer: ```suggestion
Inter-Module Communication framework for secure module authorization and OCAPs.
```

> framework. When implemented, this could also serve as an alternative to the existing paradigm of passing keepers between

Reviewer: ```suggestion
When implemented, this could also serve as an alternative to the existing paradigm of passing keepers between
```

In [ADR 021](./adr-021-protobuf-query-encoding.md), a mechanism for using protobuf service definitions to define queriers
was introduced and in [ADR 31](./adr-031-msg-service.md), a mechanism for using protobuf service to define `Msg`s was added.
Protobuf service definitions generate two golang interfaces representing the client and server sides of a service plus

> some helper code. Here is a minimal example for the bank `x/bank.MsgSend` message type:

Reviewer: ```suggestion
some helper code. Here is a minimal example for the bank `cosmos.bank.Msg/Send` message type:
```

evolution.
2. The separation between the client and server interfaces will allow us to insert permission checking code in between
the two which checks if one module is authorized to send the specified `Msg` to the other module providing a proper
object capability system (see below).
3. The router for inter-module communication gives us a convenient place to handle rollback of transactions,
enabling atomicy of operations ([currently a problem](https://github.com/cosmos/cosmos-sdk/issues/8030)). Any failure within a module-to-module call would result in a failure of the entire
transaction

This mechanism has the added benefits of:
- reducing boilerplate through code generation, and
- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC

### Inter-module Communication

To use the `Client` generated by the protobuf compiler we need a `grpc.ClientConn` [interface](https://github.com/regen-network/protobuf/blob/cosmos/grpc/types.go#L12)
implementation. For this we introduce
a new type, `ModuleKey`, which implements the `grpc.ClientConn` interface. `ModuleKey` can be thought of as the "private

> key" corresponding to a module account, where authentication is provided through use of a special `Invoker()` function (gRPC middleware),

Reviewer: I don't think it's really quite "middleware"

Author: right, it's not the best framing. I will remove that.

described in more detail below.

Blockchain users (external clients) use their account's private key to sign transactions containing `Msg`s where they are listed as signers (each
message specifies required signers with `Msg.GetSigner`). The authentication checks is performed by `AnteHandler`.

Here, we extend this process, by allowing modules to be identified in `Msg.GetSigners`. When a module creates a `Msg` for another module,
it's `ModuleKey` is used to send (through the `ClientConn` interface we describe below) and is set as a sole "signer". It's worth to note
that we don't store any cryptography signature in this case.
For example, module `A` could use it's `A.ModuleKey` to create `MsgSend` object for `/cosmos.bank.Msg/Send` transaction. `MsgSend` validation
will assure that the `from` account (`A.ModuleKey` in this case) is signer.

Here's an example of a hypothetical module `foo` interacting with `x/bank`:
```go
package foo

type FooMsgServer {
  // ...

  bankQuery bank.QueryClient
  bankMsg   bank.MsgClient
}

> func NewFooMsgServer(...) MsgServer {

Reviewer: `moduleKey` should probably be a constructor param

Author: Yes, good idea :+1:

key" corresponding to a module account, where authentication is provided through use of a special `Invoker()` function,
described in more detail below.

Blockchain users (external clients) use their account's private key to sign transactions containing `Msg`s where they are listed as signers (each
message specifies required signers with `Msg.GetSigner`). The authentication checks is performed by `AnteHandler`.

Here, we extend this process, by allowing modules to be identified in `Msg.GetSigners`. When a module creates a `Msg` for another module,
it's `ModuleKey` is used to send (through the `ClientConn` interface we describe below) and is set as a sole "signer". It's worth to note
that we don't store any cryptography signature in this case.
For example, module `A` could use it's `A.ModuleKey` to create `MsgSend` object for `/cosmos.bank.Msg/Send` transaction. `MsgSend` validation
will assure that the `from` account (`A.ModuleKey` in this case) is signer.

Here's an example of a hypothetical module `foo` interacting with `x/bank`:
```go
package foo

type FooMsgServer {
  // ...

  bankQuery bank.QueryClient
  bankMsg   bank.MsgClient
}

> func NewFooMsgServer(moduleKey, ...) FooMsgServer {

Reviewer: ```suggestion
func NewFooMsgServer(moduleKey RootModuleKey, ...) FooMsgServer {
```

The two `ModuleKey` types are `RootModuleKey` and `DerivedModuleKey`:

```go
type Invoker func(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error

> func Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error

Reviewer 2: Looks like this got taken care of.

As an example, x/slashing's Slash must call x/staking's Slash, but we don't want to expose x/staking's Slash to end users
and clients.

> Internal protobuf services will be defined in a corresponding `internal.proto` files in the given module's

Reviewer: ```suggestion
Internal protobuf services will be defined in a corresponding `internal.proto` file in the given module's
```
